internal/enginenetx/httpsdialerstats.go

```diff
@@ -0,0 +1,314 @@
+	mu sync.Mutex
+
+	// root is the root container for stats
+	root *HTTPSDialerStatsRootContainer
+}
+
+// HTTPSDialerStatsKey is the key used in the key-value store to access the state.
+const HTTPSDialerStatsKey = "httpsdialerstats.state"
+
+// errDialerStatsContainerWrongVersion means that the stats container document has the wrong version number.
+var errDialerStatsContainerWrongVersion = errors.New("wrong stats container version")
+
+// loadHTTPSDialerStatsRootContainer loads a state container from the given key-value store.
+func loadHTTPSDialerStatsRootContainer(kvStore model.KeyValueStore) (*HTTPSDialerStatsRootContainer, error) {
+	// load data from disk
+	data, err := kvStore.Get(HTTPSDialerStatsKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// parse as JSON
+	var container HTTPSDialerStatsRootContainer
+	if err := json.Unmarshal(data, &container); err != nil {
+		return nil, err
+	}
+
+	// make sure the version is OK
+	if container.Version != HTTPSDialerStatsContainerVersion {
+		err := fmt.Errorf(
+			"%s: %w: expected=%d got=%d",
+			HTTPSDialerStatsKey,
+			errDialerStatsContainerWrongVersion,
+			HTTPSDialerStatsContainerVersion,
+			container.Version,
+		)
+		return nil, err
+	}
+
+	return &container, nil
+}
+
+// NewHTTPSDialerStatsManager constructs a new instance of [*HTTPSDialerStatsManager].
+func NewHTTPSDialerStatsManager(kvStore model.KeyValueStore, logger model.Logger) *HTTPSDialerStatsManager {
+	root, err := loadHTTPSDialerStatsRootContainer(kvStore)
+	if err != nil {
+		root = NewHTTPSDialerStatsRootContainer()
+	}
+
+	return &HTTPSDialerStatsManager{
+		TimeNow: time.Now,
+		root:    root,
+		kvStore: kvStore,
+		logger:  logger,
+		mu:      sync.Mutex{},
+	}
+}
+
+var _ HTTPSDialerStatsTracker = &HTTPSDialerStatsManager{}
+
+// OnStarting implements HTTPSDialerStatsManager.
+func (mt *HTTPSDialerStatsManager) OnStarting(tactic *HTTPSDialerTactic) {
+	// get exclusive access
+	defer mt.mu.Unlock()
+	mt.mu.Lock()
+
+	// get the record
+	record, found := mt.root.GetLocked(tactic)
+	if !found {
+		record = &HTTPSDialerStatsTacticRecord{
+			CountStarted:              0,
+			CountTCPConnectError:      0,
+			CountTLSHandshakeError:    0,
+			CountTLSVerificationError: 0,
+			CountSuccess:              0,
+			HistoTCPConnectError:      map[string]int64{},
+			HistoTLSHandshakeError:    map[string]int64{},
+			HistoTLSVerificationError: map[string]int64{},
+			LastUpdated:               time.Time{},
+			Tactic:                    tactic.Clone(), // avoid storing the original
+		}
+		mt.root.SetLocked(tactic, record)
+	}
+
+	// update stats
+	record.CountStarted++
+	record.LastUpdated = mt.TimeNow()
+}
+
+// OnTCPConnectError implements HTTPSDialerStatsManager.
+func (mt *HTTPSDialerStatsManager) OnTCPConnectError(ctx context.Context, tactic *HTTPSDialerTactic, err error) {
+	// avoid updating stats if the operation has been interrupted
+	if ctx.Err() != nil {
+		return
```

Comment:
Maybe we should have an "interrupted" counter here, otherwise we would be missing some number of operations from the overall statistics, which would confuse people looking at the data.